### ISSUE TITLE
getaddrinfo: Name or service not known for sloppy.io hosting

### ISSUE THREAD

**Issue by JannikZed** [mattermost/mattermost-docker#142]
Hello there,
I'm currently trying to get mattermost in docker running for a docker hosting-provider called sloppy.io. I managed it to connect app with db, web with app, but app is not answering requests at port 80... in the logs I see this entry every second: mattermost services app nc: getaddrinfo: Name or service not known
I don't know, which mattermost-service is causing this issue, but it would be nice, if someone could give me some "inspiration" what is causing this issue. 

Thank you

**Comment by pichouk** [comment]
I don't know how [sloppy.io](https://sloppy.io) works, could you provide the configuration you pushed into it and, if possible, the entire log on Mattermost app ?

**Comment by JannikZed** [comment]
Sloppy is Apache Mesos based. The difference from docker compose is, that you need more "fully qualified domain names" to address the containers. For example: At the web container I had to change proxy_pass http://app:80; to proxy_pass http://app.services.mattermost.jannikzinkl:80;
Here is the complete log I received: 

2017-06-16 15:31:28 mattermost services app No configuration file /mattermost/config/config.json
2017-06-16 15:31:28 mattermost services app Creating a new one
2017-06-16 15:31:28 mattermost services app Configure database connection...OK
2017-06-16 15:31:29 mattermost services app nc: getaddrinfo: Name or service not known
2017-06-16 15:31:30 mattermost services app nc: getaddrinfo: Name or service not known
2017-06-16 15:31:32 mattermost services app nc: getaddrinfo: Name or service not known
2017-06-16 15:31:33 mattermost services app nc: getaddrinfo: Name or service not known
2017-06-16 15:31:34 mattermost services app nc: getaddrinfo: Name or service not known
2017-06-16 15:31:35 mattermost services app nc: getaddrinfo: Name or service not known
2017-06-16 15:31:36 mattermost services app nc: getaddrinfo: Name or service not known
2017-06-16 15:31:39 mattermost services app nc: getaddrinfo: Name or service not known
2017-06-16 15:31:40 mattermost services app nc: getaddrinfo: Name or service not known

That's all that is happening..
The data I insert is kind of the same than the docker-compose file out of this repo. This there a possibility to activate a debug-mode via environment variables? 
Do you have an idea, at which stage of the starting-process this strange "getaddrinfo: Name or service not known" is coming?

**Comment by pichouk** [comment]
You might have an issue with your database connection, it seems to failed on the [app's `docker-entry.sh`](https://github.com/mattermost/mattermost-docker/blob/master/app/docker-entry.sh) file (see the loop line `58`).
Could you modify this file and add this on line `57` please ? 
```
echo $DB_HOST
echo $DB_PORT_NUMBER
```
I think that `DB_HOST` variable is not correct.

**Comment by nagylaszlo77** [comment]
Helo! I have a same problem with same error message. But i think the problem start at web app and it restart again and again. I can enter the app container but the web not. 
I tried to describe this here:
https://forum.mattermost.org/t/mattermost-start-problem/3408

There are all error message.

**Comment by JannikZed** [comment]
Hi,I'm in vacation right now and i will update this issue with the insights I got until now from my environment on Wednesday.
	





On Mon, Jun 19, 2017 at 9:21 AM +0200, "nagylaszlo77" <notifications@github.com> wrote:












Helo! I have a same problem with same error message. But i think the problem start at web app and it restart again and again. I can enter the app container but the web not.

I tried to describe this here:

https://forum.mattermost.org/t/mattermost-start-problem/3408


There are all error message.

**Comment by pichouk** [comment]
Thanks for feedback. I think that your containers are not connected correctly. I don't know if you provide a `docker-compose.yml` file to `sloppy.io` but it seems that your app container cannot resolve `db` hostname and your web container cannot resolve the `app` hostname.

Can you dump the exact configuration and/or commands you use to run those containers please (just obfuscate passwords) ?

**Comment by nagylaszlo77** [comment]
This is the yml file:

```
version: '2'
services:
  mattermost_app:
    image: 'mattermost/mattermost-prod-app'
    restart: always
    volumes:
      - './mattermost/app/config:/mattermost/config'
      - './mattermost/app/data:/mattermost/data'
      - './mattermost/app/logs:/mattermost/logs'
      - '/etc/localtime:/etc/localtime:ro'
    environment:
      MM_USERNAME: mmuser
      MM_PASSWORD: #####_#####
      MM_DBNAME: mattermost
    depends_on:
      - mattermost_db
 mattermost_db:
    image: 'mattermost/mattermost-prod-db'
    hostname: 'mattermost-db'
    restart: always
    volumes:
      - './mattermost/db/data:/var/lib/postgresql/data'
      - '/etc/localtime:/etc/localtime:ro'
    environment:
      POSTGRES_USER: mmuser
      POSTGRES_PASSWORD: #####_#####
      POSTGRES_DB: mattermost
 mattermost_web:
    image: 'mattermost/mattermost-prod-web'
    ports:
      - '81:80'
      - '444:443'
    restart: always
    volumes:
      - './mattermost/web/cert:/cert:ro'
      - '/etc/localtime:/etc/localtime:ro'
    environment:
      VIRTUAL_HOST: mattermost.local
    depends_on:
      - mattermost_app

```

**Comment by pichouk** [comment]
Ok this is what I thought. Try with this one : 
```
version: '2'
services:
  mattermost_app:
    image: 'mattermost/mattermost-prod-app'
    restart: always
    volumes:
      - './mattermost/app/config:/mattermost/config'
      - './mattermost/app/data:/mattermost/data'
      - './mattermost/app/logs:/mattermost/logs'
      - '/etc/localtime:/etc/localtime:ro'
    environment:
      MM_USERNAME: mmuser
      MM_PASSWORD: #####_#####
      MM_DBNAME: mattermost
    link:
      - mattermost_db:db
 mattermost_db:
    image: 'mattermost/mattermost-prod-db'
    hostname: 'mattermost-db'
    restart: always
    volumes:
      - './mattermost/db/data:/var/lib/postgresql/data'
      - '/etc/localtime:/etc/localtime:ro'
    environment:
      POSTGRES_USER: mmuser
      POSTGRES_PASSWORD: #####_#####
      POSTGRES_DB: mattermost
 mattermost_web:
    image: 'mattermost/mattermost-prod-web'
    ports:
      - '81:80'
      - '444:443'
    restart: always
    volumes:
      - './mattermost/web/cert:/cert:ro'
      - '/etc/localtime:/etc/localtime:ro'
    environment:
      VIRTUAL_HOST: mattermost.local
    link:
      - mattermost_app:app
```
(The only difference is on the `link` instruction replacing the `depends_on`).

I didn't notice that our `docker-compose.yml` example used `depends_on`. I will update this.

**Comment by pichouk** [comment]
(I made a mistake on my configuration, I edit my previous message. Just posting to trigger notifications)

**Comment by nagylaszlo77** [comment]
ok its a good solution of this problem just nut link -> links is a correct

**Comment by pichouk** [comment]
Yeah I made a mistake... So is it solving the issue for you ?

@JannikZed I'll wait for your feedback, to be sure it was the same issue for you. When you will have test, and if it works, you could close.

**Comment by nagylaszlo77** [comment]
Yes thanks a lot! Now have a little db error but with curl i get the HTML.

**Comment by pichouk** [comment]
Feel free to open another issue if you're stuck with it.

**Comment by nagylaszlo77** [comment]
Ok, thanks for your help!

**Comment by JannikZed** [comment]
Hello everybody. 
Thank you @pichouk for pointing out the solution to set "$DB_HOST". Wouldn't it be a good idea to describe this environment-variable also in the realm-file? 

The networking seems to work now. Still the service is not running for me :( First it looks all good:
2017-06-16 19:45:32 mattermost services app OK
2017-06-16 19:45:32 mattermost services app Wait until database db.services.mattermost.jannikzinkl.node.intern:5432 is ready...
2017-06-16 19:45:32 mattermost services app Starting platform

After some time I see this:
2017-06-22 06:38:38 mattermost services web 2017/06/22 06:38:14 [emerg] 13#13: host not found in upstream "app.services.mattermost.jannikzinkl" in /etc/nginx/conf.d/mattermost.conf:21

And the App-Container is restarting after some time without writing any new log-entry..

**Comment by pichouk** [comment]
The `DB_HOST` is already mentioned in README but I agree that there is some important environment variables we should explain better, and some hardcoded value we should fix. I'll create some tasks for this.

It seems that your `web` container cannot resolve your `app` container now.
- Did you modify the nginx configuration file for the `web` image ? If you can enter in your `web` container, could you `cat /etc/nginx/conf.d/mattermost.conf` ?
- What is the command/docker-compose file you use to run the `web` container ? The `web` container should have a mapping from hostname `app` to the real hostname of your `app` container.

**Comment by JannikZed** [comment]
Is there another place where I have to change the hostname `app` than in the mattermost-file at  proxy_pass http://app:80;?

I'm close to giving it up :( I don't get it running and I don't have a lot of possibilities to debug it. Nothing seems to work

**Comment by pichouk** [comment]
I don't understand : is your problem on your `web` or `app` container ? You previously show a log entry with the starting platform message ( `2017-06-16 19:45:32 mattermost services app Starting platform`).

I don't know how sloppy.io works. What kind of configuration did you provide ? The docker-compose file ?
Do you have a console access to your containers ?

**Comment by pichouk** [comment]
Ok I just create an account on sloppy.io to understand how it works.
For things, it seems that you need to provide a docker image. What images did you provide for your `app` container ? [The public one on Docker Hub](https://hub.docker.com/r/mattermost/mattermost-prod-app/) is Mattermost Entreprise edition.
I don't understand how you can link containers with sloppy.io. Do you know if it is possible ?

**Comment by JannikZed** [comment]
That's the strange thing ... Sometimes it looks like the "app" container is running and sometimes not :( 
Sloppy connects all backend-services with each other automatically. You can use domain-names to connect each-other. Unfortunately I don't get access to the docker-container, that makes it so difficult. I build up a small CI-pipeline which is building the containers locally with docker-compose build and pushes them to gitlab registry afterwards. Sloppy is deploying from there. 
You control sloppy with a sloppy.yml-file which looks similar to the docker-compose file but not completely:
````yaml
version: "v1"
project: "mattermost"
services:
  services:
      db:
        image: $db
        instances: 1
        mem: 64
        volumes:
          - path: "/var/lib/postgresql/data"
            size: "8GB"
        env:
          - POSTGRES_USER: "mmuser"
          - POSTGRES_PASSWORD: "XXXX"
          - POSTGRES_DB: "mm_DB"
        # uncomment the following to enable backup
        #  - AWS_ACCESS_KEY_ID=XXXX
        #  - AWS_SECRET_ACCESS_KEY=XXXX
        #  - WALE_S3_PREFIX=s3://BUCKET_NAME/PATH
        #  - AWS_REGION=us-east-1
        #  in case your config is not in default location
        #  - MM_CONFIG=/mattermost/config/config.jso
      app:
        image: $app
        instances: 1
        mem: 128
        volumes:
          - path: "/mattermost/data"
            size: "8GB"
        env:
        # set same as db environment
          - MM_USERNAME: "mmuser"
          - MM_PASSWORD: "XXXX"
          - MM_DBNAME: "mm_DB"
          - MM_CONFIG: "/mattermost/data/config.json"
          - DB_HOST: "db.services.mattermost"
        dependencies:
          - "db"
      web:
        image: $web
        instances: 1
        mem: 64
        ports: 80
        domain: "triebwork-mm.sloppy.zone"
        env:
          - MATTERMOST_ENABLE_SSL: "false"
        dependencies:
          - "app"
````

I think that the network connectivity between app and database is running, as I can see this: `echo "Wait until database $DB_HOST:$DB_PORT_NUMBER is ready..."`and afterwards it's jumping directly out of the loop and is not entering the "until"-part.

Thank you for going so deep into helping me! It's working for me locally with docker-compose but seems to be hard to get it running without compose.

**Comment by pichouk** [comment]
I didn't find how to push a sloppy.yml file, maybe it is not on the free trial, so it is quite difficult for me to reproduce...
I don't know why the app container sometimes start, sometimes not. I don't understand the value of the `DB_HOST` variable (`db.services.mattermost`) : how the app container could resolve this ?

For the web container, a possible workaround, if you have access to your container logs, is to add a lot of `echo` in your entrypoint files,  for each environment variables/hostname involved here.I don't know how the `dependencies` is process by Sloppy.io but it seems to be an issue with the connection between the web container and the app container.

**Comment by xcompass** [comment]
@JannikZed it seems your app and db container are fine, but web container couldn't resolve upstream hosts (app container(s)). Just did some dig on sloppy.io. It seems all apps don't have dedicated IP and they provide a load balancer to redirect traffic (https://sloppy.io/knowledge-base/domains/, bottom the page). In that case, you can skip the web container, which acts as a load balancer/proxy when you run docker-compose. You just need app and db container. Then setup sloppy LB to redirect requests to your app container directly. You can also run multiple app containers for horizontal scaling. Hope it helps.

**Comment by JannikZed** [comment]
thank you for all your feedback! I think I'm getting closer to the root-cause. I consulted sloppy customer-care and they told me, that they are blocking kernel-tools like ping and nectat! 
Furthermore I activated logging to the stdout and I see the following:  [2017/06/26 10:22:19 UTC] [INFO] Pinging SQL master database
Is mattermost app-container using ping-tools? Might this being blocked by sloppy be the root-cause?

**Comment by pichouk** [comment]
`app` container use netcat to check if the database is alive a startup [as you can see here](https://github.com/mattermost/mattermost-docker/blob/master/app/docker-entry.sh#L59)
I don't know how to check that the database connection is OK without using `nc`. You can try to remove the `until` loop and just sleep for few seconds. It's less secure I guess it could work.

**Comment by xcompass** [comment]
But in your log above
```
2017-06-16 19:45:32 mattermost services app OK
2017-06-16 19:45:32 mattermost services app Wait until database db.services.mattermost.jannikzinkl.node.intern:5432 is ready...
2017-06-16 19:45:32 mattermost services app Starting platform
```

It seems that `nc` successfully detected the database. 

Also the issue related to the web container should not be affected by the blocking of the kernel-tools as it doesn't rely on them. 

Anyway, as @pichouk suggested, If sloopy support auto restart containers when they exit, you can remove `nc` and set the sleep longer. You should be fine.

**Comment by JannikZed** [comment]
Hi,
Yes, I already deleted the netcat part and that help me go a little further. I added this here, just to see if the network connectivity is working: cat </dev/tcp/app.services.mattermost.jannikzinkl/5432
And it does!
Nevertheless the last thing I see before app crashes is this: 
05 mattermost services app [2017/06/27 12:40:03 UTC] [INFO] Enterprise Enabled: false
2017-06-27 12:40:05 mattermost services app [2017/06/27 12:40:03 UTC] [INFO] Current working directory is /mattermost/bin
2017-06-27 12:40:05 mattermost services app [2017/06/27 12:40:03 UTC] [INFO] Loaded config file from /mattermost/config/config.json
2017-06-27 12:40:05 mattermost services app [2017/06/27 12:40:03 UTC] [INFO] Server is initializing...
2017-06-27 12:40:05 mattermost services app [2017/06/27 12:40:03 UTC] [INFO] Pinging SQL master database

Afterwards I see in sloppy: Container exited with status 102. No luck for me :/

**Comment by pichouk** [comment]
Not sure but maybe it's because the database is not startup yet (that's why the loop on `nc` was useful). You could try to add a `sleep 10` just to see if it help.

**Comment by xcompass** [comment]
And also could you post the db container log along with app log as well with the same run? I want to make sure the ping happened after db started. Exit code 102 means mattermost platform couldn't ping the database. Either db is not started/finished to start or there is some problem with network between two .

**Comment by JannikZed** [comment]
I added sleep 10, I restarted manually.. I even created a new container "[adminer](https://github.com/clue/docker-adminer)" on the same network. I was able to connect to the db with the same domain-name as used in the app container.. So the DB is running and reachable. 
This is the moment I give up. Can you recommend me another docker-hosting provider? Preferable in germany.

**Comment by xcompass** [comment]
@JannikZed I was able to get it working with the following sloppy.yml
```yaml
version: "v1"
project: "mattermost"
services:
  services:
    db:
      image: mattermost/mattermost-prod-db
      instances: 1
      mem: 256
      volumes:
        - path: "/var/lib/postgresql/data"
          size: "8GB"
      env:
        - POSTGRES_USER: "mmuser"
        - POSTGRES_PASSWORD: "XXXX"
        - POSTGRES_DB: "mm_DB"
      # uncomment the following to enable backup
      #  - AWS_ACCESS_KEY_ID=XXXX
      #  - AWS_SECRET_ACCESS_KEY=XXXX
      #  - WALE_S3_PREFIX=s3://BUCKET_NAME/PATH
      #  - AWS_REGION=us-east-1
      #  in case your config is not in default location
      #  - MM_CONFIG=/mattermost/config/config.jso
    app:
      image: mattermost/mattermost-prod-app
      instances: 1
      mem: 512
      volumes:
        - path: "/mattermost/data"
          size: "8GB"
      ports: 80
      domain: "mmtest.sloppy.zone"
      env:
      # set same as db environment
        - MM_USERNAME: "mmuser"
        - MM_PASSWORD: "XXXX"
        - MM_DBNAME: "mm_DB"
        - MM_CONFIG: "/mattermost/data/config.json"
        - DB_HOST: "db.services.mattermost"
      dependencies:
        - "db"
```
Here are changed I made from your yaml:
* Removed web container. Sloppy provides a load balancer, so you don't need the web container as I mentioned in the previous comment. I found one issue though: you can't scale up app container as it has a volume mounted. It seems they don't support shared volume. Even if you are able to scale up app container, mattermost won't work because it needs to access the shared volume for uploaded files and other assets on the volume. 
* Increased the mem size. I found that the size you give to db container is too small, db container stuck at initialization stage (or maybe it needs very long time to finish). I also increased app mem size to be safe. I didn't test the minimal size required.
* Moved `port` and `domain` from web to app as it is required to expose them to load balancer so that it is accessible from outside.

If you really want to run web container, you need to change this line https://github.com/mattermost/mattermost-docker/blob/master/web/mattermost#L37, from `proxy_pass http://app:80;` to `proxy_pass http://app.services.mattermost.YOUR_ID.node.intern:80;`. Similar to what you did in the previous comments but with additional `.node.intern`.

Based on my testing, I think sloppy.io is very simple and easy to use, but missing some key features, like shared volume, which usually required for app container when scaling up. So if you want to deploy high availability apps, sloppy may not be a good choice. You can want to take a look at kubernetes and it's distributions (openshift origin, deis)

@pichouk you don't need to push yaml file, but need the CLI to run it:
```bash
sloppy start sloppy.yml
```

Hope it helps.

**Comment by xcompass** [comment]
Also I was able to use the official/auto-built app/db images without any modification. So it means netcat works fine on sloppy.io.  However, it seems sloppy often swallows the last line of the log, so when you see `Wait until database db.services.mattermost.xcompass.node.intern:5432 is ready...`, it is actually passed the db checking and suppose to print `Starting platform`, but it was swallowed.  If I add another echo after the last one, `Starting platform` will actually be printed. The same for db container. The last line `database is ready for connection` is swallowed so it seems db didn't finish initialization.

**Comment by JannikZed** [comment]
Okay crazy! Finally after a lot of trying and trying I get it to work, too! I really don't know, why the application container is not starting from time to time. Is it maybe possible to change the connect_timeout to the database to a higher value or something like this? I don't know why there were so many struggles with sloppy.. Thank you for all your time and help!

Furthermore, I recognized, that the value of `MM_CONFIG=/mattermost/config/config.json`is hardcoded. Should we change it to something like this?: `MM_CONFIG=${MM_CONFIG:-/mattermost/config/config.json}`

**Comment by pichouk** [comment]
Great !
Good point about the hardcoded `MM_CONFIG`, I just commit your suggestion https://github.com/mattermost/mattermost-docker/commit/7590cefb533aa350a715a13c3b9ca3d183a73a5c

If it is solved for you, you can close :)

**Comment by JannikZed** [comment]
Again thank you for all your time!

**Comment by localsnet** [comment]
Hello colleagues!
I've just receive the same error
 `nc: getaddrinfo: Name or service not known`

 Here it is my docker-compose.file  :

`version: "2.1"
services:
  db:
    image: mysql:5.7
    restart: unless-stopped
    volumes:
      - /data/mattermost:/data/mattermost
      - ./db:/var/lib/mysql
      - /etc/localtime:/etc/localtime:ro
    environment:
      - MYSQL_USER=mmuser
      - MYSQL_PASSWORD=abc
      - MYSQL_ROOT_PASSWORD=abc
      - MYSQL_DATABASE=mattermost
    ports:
      - "3307:3306"
    network_mode: bridge
    #dns: 

  app:
    image: mattermost/mattermost-prod-app:3.9.0
    restart: unless-stopped
    volumes:
      - ./app/config:/mattermost/config:rw
      - ./app/data:/mattermost/data:rw
      - ./app/logs:/mattermost/logs:rw
      - /etc/localtime:/etc/localtime:ro  
  #  - ./data/hosts:/etc/hosts
    environment:
      - MM_USERNAME=mmuser
      - MM_PASSWORD=abc
      - MM_DBNAME=mattermost
      - DB_PORT_5432_TCP_PORT=3306
      - DB_HOST=mattermost_db_1
    ports:
      - "8066:8065"
    network_mode: bridge
    sysctls: 
      - net.ipv4.ip_forward=1
`
The line #  - ./data/hosts:/etc/hosts  I was commented, because it even didn't run an app container.
Also, in my file is lack of web part, could I use it from official mattermost docker-compose file?
Could you direct me how to connect an app with db and to add web service?
Thank you in advance.

**Comment by pichouk** [comment]
@localsnet If you are not using Sloppy.io, it may be a different issue and you should open an new Github issue, with your config (be careful, your markdown seems to have failed).
But to help you on this, this message is clearly a network issue and I see some instructions (`network_mode` and `net.ipv4.ip_forward=1`) that I don't know. Maybe you could explain why you add those :)

**Comment by localsnet** [comment]
@pichouk Thank you indeed for your quick feedback! You are right , I'm not using this one.
In fact I've been received this existing file and I need to make it to work well, so I'm trying to learn Docker on the fly, by reading documentation.

**Comment by xcompass** [comment]
@localsnet containers can not  resolve each other by name when using bridge mode as in docker doc indicated
> Docker does not support automatic service discovery on the default bridge network. If you want containers to be able to resolve IP addresses by container name, you should use user-defined networks instead.

Not sure why you added that. You can either remove it to use docker compose created user-defined network or you have to change entrypoint.sh to use dB container IP.
